io/directory.go

```diff
@@ -309,15 +350,31 @@ var _ Directory = (*UpgradeableDirectory)(nil)
 // AddChild implements the `Directory` interface. We check when adding new entries
 // if we should switch to HAMTDirectory according to global option(s).
 func (d *UpgradeableDirectory) AddChild(ctx context.Context, name string, nd ipld.Node) error {
-	if UseHAMTSharding {
-		if basicDir, ok := d.Directory.(*BasicDirectory); ok {
-			hamtDir, err := basicDir.SwitchToSharding(ctx)
-			if err != nil {
-				return err
-			}
-			d.Directory = hamtDir
+	err := d.Directory.AddChild(ctx, name, nd)
+	if err != nil {
+		return err
+	}
+
+	// Evaluate possible HAMT upgrade.
+	if HAMTShardingSize == 0 {
+		return nil
+	}
+	basicDir, ok := d.Directory.(*BasicDirectory)
+	if !ok {
+		return nil
+	}
+	if basicDir.estimatedSize >= HAMTShardingSize {
+		// FIXME: Ideally to minimize performance we should check if this last
+		//  `AddChild` call would bring the directory size over the threshold
+		//  *before* executing it since we would end up switching anyway and
+		//  that call would be "wasted". This is a minimal performance impact
+		//  and we prioritize a simple code base.
```

Comment:
Completely agree. I'd leave this comment but remove the "FIXME", we shouldn't fix this.